pcmanfm/desktopwindow.cpp

```diff
@@ -633,39 +636,119 @@ void DesktopWindow::updateWallpaper() {
+                        if(w_ratio > h_ratio && wallpaperMode_ == WallpaperFit){
+                            //fit vertical, space on sides.
+                            scaled = image.scaledToHeight(scr->geometry().height(), Qt::SmoothTransformation);
+                            int x = scr->geometry().x() + (scr->geometry().width() - scaled.width()) / 2;
+                            int y = scr->geometry().y();
+                            painter.drawImage(x, y, scaled);
+                        }
+                        //zoom
+                        if(w_ratio <= h_ratio && wallpaperMode_ == WallpaperZoom){
+                            //fit vertical, sides leftovers
+                            scaled = image.scaledToHeight(scr->geometry().height(), Qt::SmoothTransformation);
+                            //crop to avoid superposition
+                            int x_gap = (scaled.width() - scr->geometry().width()) / 2;
+                            scaled = scaled.copy(x_gap, 0, scaled.width(), scaled.height());
+                            int x = scr->geometry().x();
+                            int y = scr->geometry().y();
+                            painter.drawImage(x, y, scaled);
+                        }
+                        if(w_ratio > h_ratio && wallpaperMode_ == WallpaperZoom){
+                           //fit horizonatal, up and down leftovers
+                            scaled = image.scaledToWidth(scr->geometry().width(), Qt::SmoothTransformation);
+                            //crop to avoid superposition
+                            int y_gap = (scaled.height() - scr->geometry().height()) / 2;
+                            scaled = scaled.copy(0, y_gap, scaled.width(), scaled.height());
+                            int x = scr->geometry().x();
+                            int y = scr->geometry().y();
+                            painter.drawImage(x, y, scaled);
+                        }
+                    }
                 }
             }
-            if(!image.isNull()) {
-                pixmap = QPixmap(size());
-                QPainter painter(&pixmap);
-                pixmap.fill(bgColor_);
-                int x = (width() - image.width()) / 2;
-                int y = (height() - image.height()) / 2;
-                painter.drawImage(x, y, image);
+            else {
+                if(wallpaperMode_ == WallpaperCenter) {
+                    image = getWallpaperImage(); // load original image
+                }
+                else if(wallpaperMode_ == WallpaperFit || wallpaperMode_ == WallpaperZoom) {
+                    // calculate the desired size
+                    QImageReader reader(wallpaperFile_);
+                    QSize origSize = reader.size(); // get the size of the original file
+                    if(reader.transformation() & QImageIOHandler::TransformationRotate90) {
+                        Settings& settings = static_cast<Application* >(qApp)->settings();
```

Comment:
@hmollercl 
Would you please remove the above line -- `settings` is already  defined -- and also include https://github.com/lxqt/pcmanfm-qt/pull/1045 in your PR (it's about adding a nullity check)? In that way, you won't need to re-base your PR and I'll merge it after testing it again.

Reply:
removed that re-definition of settings and added #1045 code.